cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -332,9 +336,8 @@ private void processMetrics() {
         try {
           Deque<MetricValues> recordsCopy = new LinkedList<>();
           Map<TopicIdMetaKey, byte[]> topicMessageIdsCopy = new HashMap<>(topicMessageIds);
-          // TODO: (CDAP-8327) there is a risk of running out-of-memory if other threads keep writing to records
           Iterator<MetricValues> iterator = records.iterator();
-          while (iterator.hasNext()) {
+          while (iterator.hasNext() && recordsCopy.size() < persistThreshold) {
```

Comment:
If you terminate earlier, that means the unprocessed entries will be left in the shared queue, which potentially it can grow faster than the `persistThreshold` could empty it, causing OOM